[tesla] Exception occurred - communication_error

After the [fix of a recent API token problem](https://github.com/openhab/openhab-addons/pull/12530), another problem with the tesla binding has come up.
@tkranz gets this error

```2022-03-27 17:52:41.378 [ERROR] [internal.handler.TeslaAccountHandler] - An exception occurred while connecting to the Tesla back-end: 'null'
at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.authenticate(TeslaAccountHandler.java:259) ~[?:?]
at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.lambda$0(TeslaAccountHandler.java:377) ~[?:?]
2022-03-27 17:52:41.443 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:12345:54321' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2022-03-27 17:52:41.526 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:12345:54321' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
2022-03-27 17:52:56.379 [ERROR] [internal.handler.TeslaAccountHandler] - An exception occurred while connecting to the Tesla back-end: 'null'
at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.authenticate(TeslaAccountHandler.java:259) ~[?:?]
at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.lambda$0(TeslaAccountHandler.java:377) ~[?:?]
```

and I'm getting this error
```
2022-03-27 17:01:36.557 [WARN ] [internal.handler.TeslaVehicleHandler] - Exception occurred in slowStateRunnable
java.util.ConcurrentModificationException: null
	at java.util.ArrayList.sort(ArrayList.java:1752) ~[?:?]
	at org.apache.cxf.jaxrs.provider.ProviderFactory.sortReaders(ProviderFactory.java:734) ~[?:?]
	at org.apache.cxf.jaxrs.provider.ProviderFactory.setCommonProviders(ProviderFactory.java:662) ~[?:?]
	at org.apache.cxf.jaxrs.client.ClientProviderFactory.setProviders(ClientProviderFactory.java:73) ~[?:?]
	at org.apache.cxf.jaxrs.provider.ProviderFactory.setUserProviders(ProviderFactory.java:862) ~[?:?]
	at org.apache.cxf.jaxrs.client.spec.ClientImpl$WebTargetImpl.request(ClientImpl.java:282) ~[?:?]
	at org.apache.cxf.jaxrs.client.spec.ClientImpl$WebTargetImpl.request(ClientImpl.java:377) ~[?:?]
	at org.openhab.binding.tesla.internal.handler.TeslaVehicleHandler.queryVehicle(TeslaVehicleHandler.java:738) ~[?:?]
	at org.openhab.binding.tesla.internal.handler.TeslaVehicleHandler.queryVehicleAndUpdate(TeslaVehicleHandler.java:772) ~[?:?]
	at org.openhab.binding.tesla.internal.handler.TeslaVehicleHandler.lambda$0(TeslaVehicleHandler.java:977) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
```

Since the night, I get a COMMUNICATION_ERROR for the car thing (the account thing is online). I get this warn an error once a minute now:
```
2022-03-28 00:29:41.745 [WARN ] [internal.handler.TeslaVehicleHandler] - Got an unsuccessful result, setting vehicle to offline and will try again
2022-03-28 00:29:41.746 [ERROR] [internal.handler.TeslaVehicleHandler] - An error occurred while querying the vehicle
```

Manually disabling the account and car thing and then enabling them again, brings both back online.

Not sure if it is related, but I now get
```
08:26:10.292 [DEBUG] [.internal.handler.TeslaAccountHandler] - Setting up an authenticated connection to the Tesla back-end
08:26:10.292 [ERROR] [.internal.handler.TeslaAccountHandler] - An exception occurred while connecting to the Tesla back-end: 'null'
java.lang.NullPointerException: null
```
in my logs.

yes, same for me, account and car go offline after some time 
2022-03-30 10:17:06.458 [ERROR] [internal.handler.TeslaAccountHandler] - An exception occurred while connecting to the Tesla back-end: 'null'
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.authenticate(TeslaAccountHandler.java:259) ~[?:?]
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.lambda$0(TeslaAccountHandler.java:377) ~[?:?]
2022-03-30 10:17:18.693 [WARN ] [internal.handler.TeslaVehicleHandler] - Got an unsuccessful result, setting vehicle to offline and will try again
2022-03-30 10:17:18.693 [ERROR] [internal.handler.TeslaVehicleHandler] - An error occurred while querying the vehicle
2022-03-30 10:17:18.750 [WARN ] [internal.handler.TeslaVehicleHandler] - Got an unsuccessful result, setting vehicle to offline and will try again
2022-03-30 10:17:18.750 [ERROR] [internal.handler.TeslaVehicleHandler] - An error occurred while querying the vehicle
2022-03-30 10:17:21.459 [ERROR] [internal.handler.TeslaAccountHandler] - An exception occurred while connecting to the Tesla back-end: 'null'
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.authenticate(TeslaAccountHandler.java:259) ~[?:?]
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.lambda$0(TeslaAccountHandler.java:377) ~[?:?]
2022-03-30 10:17:36.460 [ERROR] [internal.handler.TeslaAccountHandler] - An exception occurred while connecting to the Tesla back-end: 'null'
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.authenticate(TeslaAccountHandler.java:259) ~[?:?]
	at org.openhab.binding.tesla.internal.handler.TeslaAccountHandler.lambda$0(TeslaAccountHandler.java:377) ~[?:?]
	
	
	
after disable and enable account and car go online

2022-03-30 10:18:24.147 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:5YJSA7E21HF178912' changed from OFFLINE (COMMUNICATION_ERROR) to UNINITIALIZED
2022-03-30 10:18:24.149 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:5YJSA7E21HF178912' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)
2022-03-30 10:18:24.149 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:66da4728df' changed from OFFLINE (COMMUNICATION_ERROR) to UNINITIALIZED
2022-03-30 10:18:24.152 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:66da4728df' changed from UNINITIALIZED to UNINITIALIZED (HANDLER_MISSING_ERROR)
2022-03-30 10:18:24.152 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:account:cf6c1bd88a' changed from OFFLINE (COMMUNICATION_ERROR) to UNINITIALIZED
2022-03-30 10:18:24.179 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:account:cf6c1bd88a' changed from UNINITIALIZED to UNINITIALIZED (DISABLED)
2022-03-30 10:18:24.180 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:5YJSA7E21HF178912' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to UNINITIALIZED (BRIDGE_UNINITIALIZED)
2022-03-30 10:18:24.180 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:66da4728df' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to UNINITIALIZED (BRIDGE_UNINITIALIZED)
2022-03-30 10:18:42.816 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:account:cf6c1bd88a' changed from UNINITIALIZED (DISABLED) to INITIALIZING
2022-03-30 10:18:42.820 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:account:cf6c1bd88a' changed from INITIALIZING to UNKNOWN
2022-03-30 10:18:42.820 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:66da4728df' changed from UNINITIALIZED (BRIDGE_UNINITIALIZED) to INITIALIZING
2022-03-30 10:18:42.821 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:5YJSA7E21HF178912' changed from UNINITIALIZED (BRIDGE_UNINITIALIZED) to INITIALIZING
2022-03-30 10:18:42.824 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:66da4728df' changed from INITIALIZING to UNKNOWN
2022-03-30 10:18:42.825 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:5YJSA7E21HF178912' changed from INITIALIZING to UNKNOWN
2022-03-30 10:18:43.299 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:account:cf6c1bd88a' changed from UNKNOWN to ONLINE
2022-03-30 10:18:43.850 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:5YJSA7E21HF178912' changed from UNKNOWN to ONLINE
2022-03-30 10:18:43.850 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'tesla:models:cf6c1bd88a:66da4728df' changed from UNKNOWN to ONLINE
2022-03-30 10:18:43.851 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'TeslaState' changed from asleep to online

--- a/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaAccountHandler.java
+++ b/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaAccountHandler.java
@@ -170,6 +170,12 @@ public class TeslaAccountHandler extends BaseBridgeHandler {
     protected boolean checkResponse(Response response, boolean immediatelyFail) {
         if (response != null && response.getStatus() == 200) {
             return true;
+        } else if (response != null && response.getStatus() == 401) {
+            logger.debug("The access token has expired, trying to get a new one.");
+            ThingStatusInfo authenticationResult = authenticate();
+            updateStatus(authenticationResult.getStatus(), authenticationResult.getStatusDetail(),
+                    authenticationResult.getDescription());
+            return false;
         } else {
             apiIntervalErrors++;
             if (immediatelyFail || apiIntervalErrors >= API_MAXIMUM_ERRORS_IN_INTERVAL) {
@@ -203,7 +209,7 @@ public class TeslaAccountHandler extends BaseBridgeHandler {
                     response.getStatusInfo().getReasonPhrase());
 
             if (!checkResponse(response, true)) {
-                logger.error("An error occurred while querying the vehicle");
+                logger.debug("An error occurred while querying the vehicle");
                 return null;
             }
 

--- a/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaAccountHandler.java
+++ b/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaAccountHandler.java
@@ -248,7 +248,7 @@ public class TeslaAccountHandler extends BaseBridgeHandler {
         return this.getThing().getUID().getId();
     }
 
-    private ThingStatusInfo authenticate() {
+    ThingStatusInfo authenticate() {
         TokenResponse token = logonToken;
 
         boolean hasExpired = true;

--- a/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java
+++ b/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaSSOHandler.java
@@ -14,6 +14,7 @@ package org.openhab.binding.tesla.internal.handler;
 
 import static org.openhab.binding.tesla.internal.TeslaBindingConstants.*;
 
+import java.time.Instant;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
@@ -68,6 +69,7 @@ public class TeslaSSOHandler {
             TokenResponse tokenResponse = gson.fromJson(refreshTokenResponse.trim(), TokenResponse.class);
 
             if (tokenResponse != null && tokenResponse.access_token != null && !tokenResponse.access_token.isEmpty()) {
+                tokenResponse.created_at = Instant.now().getEpochSecond();
                 return tokenResponse;
             } else {
                 logger.debug("An error occurred while exchanging SSO auth token for API access token.");

--- a/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaVehicleHandler.java
+++ b/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaVehicleHandler.java
@@ -707,7 +707,7 @@ public class TeslaVehicleHandler extends BaseThingHandler {
         sendCommand(COMMAND_WAKE_UP, account.wakeUpTarget);
     }
 
-    protected Vehicle queryVehicle() {
+    protected synchronized Vehicle queryVehicle() {
         String authHeader = account.getAuthHeader();
 
         if (authHeader != null) {
@@ -720,7 +720,7 @@ public class TeslaVehicleHandler extends BaseThingHandler {
                         response.getStatusInfo().getReasonPhrase());
 
                 if (!checkResponse(response, true)) {
-                    logger.error("An error occurred while querying the vehicle");
+                    logger.debug("An error occurred while querying the vehicle");
                     return null;
                 }
 

--- a/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaVehicleHandler.java
+++ b/bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/handler/TeslaVehicleHandler.java
@@ -549,6 +549,10 @@ public class TeslaVehicleHandler extends BaseThingHandler {
     protected boolean checkResponse(Response response, boolean immediatelyFail) {
         if (response != null && response.getStatus() == 200) {
             return true;
+        } else if (response != null && response.getStatus() == 401) {
+            logger.debug("The access token has expired, trying to get a new one.");
+            account.authenticate();
+            return false;
         } else {
             apiIntervalErrors++;
             if (immediatelyFail || apiIntervalErrors >= TeslaAccountHandler.API_MAXIMUM_ERRORS_IN_INTERVAL) {

